activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb

```diff
@@ -5,15 +5,30 @@ module ConnectionAdapters
     module PostgreSQL
       class SchemaCreation < SchemaCreation # :nodoc:
         private
-          delegate :quoted_include_columns_for_index, to: :@conn
+          delegate :quoted_include_columns_for_index, :create_enum, to: :@conn
+
+          def visit_TableDefinition(o)
+            create_enums(o.columns)
+            super
+          end
 
           def visit_AlterTable(o)
+            create_enums(o.adds.map(&:column))
             sql = super
             sql << o.constraint_validations.map { |fk| visit_ValidateConstraint fk }.join(" ")
             sql << o.exclusion_constraint_adds.map { |con| visit_AddExclusionConstraint con }.join(" ")
             sql << o.unique_constraint_adds.map { |con| visit_AddUniqueConstraint con }.join(" ")
           end
 
+          def create_enums(columns)
+            enums_to_create = columns.select { |c| c.type == :enum && c.options[:values] }
+
+            enums_to_create.each do |c|
+              enum_type = c.options[:enum_type] || c.name
```

Comment:
We could save one iteration by moving the check inside the loop. But this code is not on a hot path so I think we should optimize for readability. And building `enums_to_create` on a separate line may be a better option from readability perspective. So totally up to you! 

```suggestion
            columns.each do |c|
              next unless c.type == :enum && c.options[:values]
               
              enum_type = c.options[:enum_type] || c.name
```

Reply:
I originally separated on another line for readability as you mentioned, but your suggestion is still plenty readable. Done!